service/storage_service.cc

```diff
@@ -969,13 +969,27 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
     }
 
     bool is_member = tmptr->is_member(endpoint);
-    // Update pending ranges after update of normal tokens immediately to avoid
-    // a race where natural endpoint was updated to contain node A, but A was
-    // not yet removed from pending endpoints
-    if (!is_member) {
-        tmptr->update_topology(endpoint, get_dc_rack_for(endpoint));
+    bool do_notify_joined = false;
+
+    if (endpoints_to_remove.contains(endpoint)) [[unlikely]] {
+        if (!owned_tokens.empty()) {
+            on_fatal_internal_error(slogger, format("endpoint={} is marked for removal but still owns {} tokens", endpoint, owned_tokens.size()));
```

Comment:
Is the code above responsible for ensuring owned_tokens is empty? That is, we are asserting an function-local invariant here, not a node-scope or cluster-scope invariant?

It's much easier to be sure that a function-local invariant indeed holds compared to a node-scope or cluster-scope invariant.

Reply:
Yes, `owned_tokens` is function local, and prepared in the code block right above this assertion.